File: dynamic_programming/java/CoinChange.java
public class CoinChange {
    public static int change(int amount, int[] coins) {
        int[] dp = new int[amount + 1];
        dp[0] = 1;

        for (int coin : coins) {
            for (int i = coin; i <= amount; i++) {
                dp[i] += dp[i-coin];
            }
        }
        return dp[amount];
    }

    public static void main(String[] args) {
        int a[] = {2, 3, 4};
        int t = 7;
        int res = change(t, a);
        System.out.println(res);
    }
}

File: dynamic_programming/python/CoinChange.py
# Time: O(m * n)
# Space: O(m * n)

def coin_change(coins, target):
    n = len(coins)
    dp = [1 if not i else 0 for i in range(target + 1)]
    for i in range(n):
        dp2 = [1]
        for j in range(1, target + 1):
            if j < coins[i]:    dp2.append(dp[j])
            else:    dp2.append(dp[j] + dp2[j - coins[i]])
        dp = dp2
    return dp[-1]

if __name__=='__main__':
    print(coin_change([2, 3, 4], 7))